analyzers/dataframe/Vertex.cc

```diff
@@ -0,0 +1,599 @@
+		double arg = TMath::Sqrt(TMath::Max(0.0, Rv*Rv - Di*Di)/(1.+2*Ci*Di));
+		cz += (cti*arg + zi)/sZi2;
+	}
+	xv(2) = cz / hz;
+
+        TVector3 result( xv(0), xv(1), xv(2));
+
+  	return result;
+}
+
+TVectorD Fillf(TVectorD par, TVectorD xin)
+{
+	//
+ 	// Decode input arrays
+   	//
+  	double D = par(0);
+	double p0 = par(1);
+	double C = par(2);
+	double z0 = par(3);
+	double ct = par(4);
+	//
+  	double x = xin(0);
+	double y = xin(1);
+	double z = xin(2);
+	double R2 = x * x + y * y;
+	double D2 = D*D;
+	double UpCD = 1.0 + C*D;
+	double Up2CD = 1.0 + 2 * C*D;
+	double Az = C*(z - z0) / ct;
+	//
+  	// Calculate constraints
+   	//
+  	TVectorD f(2);
+	f(0) = (R2 * C + UpCD*D)/Up2CD - y * TMath::Cos(p0) + x * TMath::Sin(p0);
+	f(1) = TMath::Sin(Az)*TMath::Sin(Az) - C*C*(R2 - D2) / Up2CD;
+	//
+	return f;
+}
+//
+
+
+TMatrixD FillD(TVectorD par, TVectorD xin)
+{
+	//
+ 	// Decode input arrays
+   	//
+  	double D = par(0);
+	double p0 = par(1);
+	double C = par(2);
+	double z0 = par(3);
+	double ct = par(4);
+	//
+  	double x = xin(0);
+	double y = xin(1);
+	double z = xin(2);
+	double R2 = x * x + y * y;
+	double D2 = D*D;
+	double UpCD = 1.0 + C*D;
+	double Up2CD = 1.0 + 2 * C*D;
+	double Az = C*(z - z0) / ct;
+	//
+  	// Calculate matrix elements
+   	//
+  	TMatrixD Do(2, 5); Do.Zero();
+	Do(0, 0) = 1.0 - 2 * C*(D + C*(R2 - D2)) / (Up2CD*Up2CD);	// df(0)/dD
+	Do(0, 1) = x * TMath::Cos(p0) + y * TMath::Sin(p0);			// df(0)/dphi0
+	Do(0, 2) = (R2 - D2) / (Up2CD*Up2CD);						// df(0)/dC
+	Do(0, 3) = 0.0;												// df(0)/dz0
+	Do(0, 4) = 0.0;												// df(0)/dct
+	Do(1, 0) = 2 * C*C*(C*R2 + D*UpCD) / (Up2CD*Up2CD);			// df(1)/dD
+	Do(1, 1) = 0.0;												// df(1)/dphi0
+	Do(1, 2) = TMath::Sin(2 * Az)*Az / C - 2 * C*(R2 - D2)*UpCD / (Up2CD*Up2CD); // df(1)/dC
+	Do(1, 3) = -TMath::Sin(2 * Az)*C / ct;						// df(1)/dz0
+	Do(1, 4) = -TMath::Sin(2 * Az)*Az / ct;						// df(1)/dct
+	//
+  	return Do;
+}
+//
+
+TMatrixD FillB(TVectorD par, TVectorD xin)
+{
+	//
+ 	// Decode input arrays
+   	//
+  	double D = par(0);
+	double p0 = par(1);
+	double C = par(2);
+	double z0 = par(3);
+	double ct = par(4);
+	//
+  	double x = xin(0);
+	double y = xin(1);
+	double z = xin(2);
+	double R2 = x * x + y * y;
+	double D2 = D*D;
+	double UpCD = 1.0 + C*D;
+	double Up2CD = 1.0 + 2 * C*D;
+	double Az = C*(z - z0) / ct;
+	//
+  	// Calculate constraints
+   	//
+  	TMatrixD B(2, 3); 
+        B.Zero();
+	B(0, 0) = 2 * C*x/Up2CD + TMath::Sin(p0);
+	B(0, 1) = 2 * C*y/Up2CD - TMath::Cos(p0);
+	B(0, 2) = 0.0;
+	B(1, 0) = -2 * x*C*C / Up2CD;
+	B(1, 1) = -2 * y*C*C / Up2CD;
+	B(1, 2) = TMath::Sin(2 * Az)*C / ct;
+	//
+  	return B;
+}
+
+
+
+
+
+edm4hep::VertexData  VertexFB( int Primary, ROOT::VecOps::RVec<edm4hep::ReconstructedParticleData> recoparticles,
+					ROOT::VecOps::RVec<edm4hep::TrackState> thetracks )
+
+{
+
+        edm4hep::VertexData result;
+
+        ROOT::VecOps::RVec<edm4hep::TrackState> tracks = getRP2TRK( recoparticles, thetracks );
```

Comment:
why calling this here and not in the python. Like the the method VertexFB only takes ```edm4hep::TrackState``` as input

Reply:
indeed, I initially coded VertexFB such that it only takes as input a collection of tracks. But then, I added the recoParticles,  in view of the fact that one may want to keep trace of the *recoParticles* that are associated with a vertex.